pkg/scheduler/partition.go

```diff
@@ -0,0 +1,1101 @@
+	return len(pc.nodes)
+}
+
+func (pc *PartitionContext) GetApplications() []*objects.Application {
+	pc.RLock()
+	defer pc.RUnlock()
+	var appList []*objects.Application
+	for _, app := range pc.applications {
+		appList = append(appList, app)
+	}
+	return appList
+}
+
+func (pc *PartitionContext) GetNodes() []*objects.Node {
+	pc.RLock()
+	defer pc.RUnlock()
+	var nodeList []*objects.Node
+	for _, node := range pc.nodes {
+		nodeList = append(nodeList, node)
+	}
+	return nodeList
+}
+
+// Add an allocation to the partition/node/application/queue during node registration.
+// Queue max allocation is not checked as the allocation is part of a new node addition.
+//
+// NOTE: this is a lock free call. It should only be called holding the Partition lock.
+func (pc *PartitionContext) addAllocation(alloc *objects.Allocation) error {
+	if pc.isStopped() {
+		return fmt.Errorf("partition %s is stopped cannot add new allocation %s", pc.Name, alloc.AllocationKey)
+	}
+
+	// if the allocation is node reported (aka recovery an existing allocation),
+	// we must not generate an UUID for it, we directly use the UUID reported by shim
+	// to track this allocation, a missing UUID is a broken allocation
+	if alloc.UUID == "" {
+		metrics.GetSchedulerMetrics().IncSchedulingError()
+		return fmt.Errorf("failing to restore allocation %s for application %s: missing UUID",
+			alloc.AllocationKey, alloc.ApplicationID)
+	}
+
+	log.Logger().Debug("adding recovered allocation",
+		zap.String("partitionName", pc.Name),
+		zap.String("appID", alloc.ApplicationID),
+		zap.String("allocKey", alloc.AllocationKey),
+		zap.String("UUID", alloc.UUID))
+
+	// Check if allocation violates any resource restriction, or allocate on a
+	// non-existent application or nodes.
+	var node *objects.Node
+	var app *objects.Application
+	var ok bool
+
+	if node, ok = pc.nodes[alloc.NodeID]; !ok {
+		metrics.GetSchedulerMetrics().IncSchedulingError()
+		return fmt.Errorf("failed to find node %s", alloc.NodeID)
+	}
+
+	if app, ok = pc.applications[alloc.ApplicationID]; !ok {
+		metrics.GetSchedulerMetrics().IncSchedulingError()
+		return fmt.Errorf("failed to find application %s", alloc.ApplicationID)
+	}
+	queue := app.GetQueue()
+
+	// check the node status again
+	if !node.IsSchedulable() {
+		metrics.GetSchedulerMetrics().IncSchedulingError()
+		return fmt.Errorf("node %s is not in schedulable state", node.NodeID)
+	}
+
+	// If the new allocation goes beyond the queue's max resource (recursive)?
+	// Only check if it is allocated not when it is node reported.
+	if err := queue.IncAllocatedResource(alloc.AllocatedResource, true); err != nil {
+		metrics.GetSchedulerMetrics().IncSchedulingError()
+		return fmt.Errorf("cannot allocate resource from application %s: %v ",
+			alloc.ApplicationID, err)
+	}
+
+	node.AddAllocation(alloc)
+	app.RecoverAllocationAsk(alloc.Ask)
+	app.AddAllocation(alloc)
+	pc.allocations[alloc.UUID] = alloc
+
+	log.Logger().Debug("recovered allocation",
```

Comment:
Is this log message correct? Shouldn't be "added allocation"?

Reply:
No, this is now only called during node registration. If a node has pre-existing allocation (recovered allocations) this will add them. This is the comment in line 925:
```
// Add an allocation to the partition/node/application/queue during node registration.
```